### ISSUE TITLE
Protobuf compiler 3.10 generates Costing_Name clashing with valhalla::Costing_Name

### ISSUE THREAD

**Issue by mloskot** [valhalla/valhalla#1970]
TL;TR: When compiling Valhalla's `.proto` files using Protobuf **3.10**, the `protoc` generates `Foo_Name` and `Foo_Parse` functions which clash with existing hand-rolled ones of the very same names (and possibly similar purpose).

Let's consider the problem using the [enum Costing in options.proto](https://github.com/valhalla/valhalla/blob/3176db5363b4a3c678532b8d496274858977776d/proto/options.proto#L33) and [Costing_Name](https://github.com/valhalla/valhalla/blob/3176db5363b4a3c678532b8d496274858977776d/src/worker.cc#L952) function for concrete example.

First, the [valhalla namespace in src/worker.cc](https://github.com/valhalla/valhalla/blob/3176db5363b4a3c678532b8d496274858977776d/src/worker.cc#L889) defines the collection of `Foo_Name` and `Foo_Parse` functions, including the `Costing_Name` function.

Second, compilation of the `options.proto` using `protoc` **3.10** produces the C++ files which contain the following

- `options.pb.h`

    ```cpp
    const std::string& Costing_Name(Costing value);

    template<typename T>
    inline const std::string& Costing_Name(T enum_t_value) {
    static_assert(::std::is_same<T, Costing>::value ||
        ::std::is_integral<T>::value,
        "Incorrect type passed to function Costing_Name.");
    return Costing_Name(static_cast<Costing>(enum_t_value));
    }
    ```

- `options.pb.cc`

    ```cpp
    const std::string& Costing_Name(
        Costing value) {
    static const bool dummy =
        ::PROTOBUF_NAMESPACE_ID::internal::InitializeEnumStrings(
            Costing_entries,
            Costing_entries_by_number,
            13, Costing_strings);
    (void) dummy;
    int idx = ::PROTOBUF_NAMESPACE_ID::internal::LookUpEnumName(
        Costing_entries,
        Costing_entries_by_number,
        13, value);
    return idx == -1 ? ::PROTOBUF_NAMESPACE_ID::internal::GetEmptyString() :
                        Costing_strings[idx].get();
    }
    ```

among other `Foo_Parse` and `Foo_Name` definitions.

This leads to the redefinition and link-error, for example, when linking tools against static archive `valhalla.lib`:

```console
valhalla.lib(worker.obj) : 
  error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl valhalla::Costing_Name(enum valhalla::Costing)" 
    (?Costing_Name@valhalla@@YAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@W4Costing@1@@Z) 
  already defined in valhalla.lib(options.pb.obj)
  [C:\projects\valhalla\build\valhalla_run_isochrone.vcxproj]
```

Compilation of `options.proto` with `protoc` **3.2** does not output any such `Foo_Parse` and `Foo_Name` definitions, so no link-error occurs.

The [valhalla-protoc-3.2-vs-3.10.zip](https://github.com/valhalla/valhalla/files/3719101/valhalla-protoc-3.2-vs-3.10.zip) contains full versions of the `protoc`-generated files for comparison:

```
valhalla-protoc-3.10-options.pb.cc
valhalla-protoc-3.10-options.pb.h
valhalla-protoc-3.2-options.pb.cc
valhalla-protoc-3.2-options.pb.h
```

My experience with Protobuf is running short to judge if this is something expected or a compatibility breakage.

------

Currently, Valhalla's CI on AppVeyor (and Azure Pipelines) configurations use the Protobuf **3.10**, but since no tools are currently built, just `valhalla.lib`, the link-error does not occur.
I am preparing PR to add `-DENABLE_TOOLS=ON` (and perhaps `-DENABLE_DATA_TOOLS=ON` too) and that attempt revealed the problem.

**Comment by kevinkreiser** [comment]
Yeah this has been annoying for a while. In previous versions of proto if the light generated stuff was missing these enum functions. It seems in proto3 they reversed that decision. Imho we can probably solve this with ifdefs and make everyone's lives easier

**Comment by kevinkreiser** [comment]
I guess we could just rename the functions as well. Which does everyone prefer?

**Comment by mloskot** [comment]
Since `worker.cc` indirectly includes some `*.pb.h` (I think `api.pb.h` for sure), then I trivial wrapping of the definitions should work, like so

```
#if GOOGLE_PROTOBUF_VERSION <= 3002000
...
#endif
```

> I guess we could just rename the functions as well. Which does everyone prefer?

If understand it correctly, none of the `protoc`-generated functions are/will be used, then renaming seems the simplest and ambiguity-free way to go.

**Comment by kevinkreiser** [comment]
@mloskot sounds good to me!

**Comment by mloskot** [comment]
@kevinkreiser Are you going to do it? If not, then I may give it a go. What name should I pick, something along these lines?

```
Costing_Name -> GetCostingName
Costing_Parse  -> ParseCosting
Options_Action_Parse -> ParseOptionsAction
Options_Action_Name -> GetOptionsActionName
```

**Comment by kevinkreiser** [comment]
I'll ship it if you PR it! I'm not at a proper computer at the moment

**Comment by mloskot** [comment]
I can PR it but later this week as I'm travelling this week.

**Comment by mloskot** [comment]
@kevinkreiser Awesome, thank you!

**Comment by mloskot** [comment]
This commit https://github.com/valhalla/valhalla/issues/1970#event-2744422126 by @kevinkreiser should also fix #1951 for @PureTryOut

**Comment by molind** [comment]
Hello everyone. I've also stumbled upon this issue. Have you noticed differences in your code vs protobuf 3.10+ code? Protobuf uses "auto_" as a string representation of Costing `auto_`. Probably this difference could break some other parts of Valhalla.

```
const std::string& Costing_Name(const odin::Costing costing) {
  static const std::string empty;
  static const std::unordered_map<int, std::string> costings{
      {odin::Costing::auto_, "auto"},
      {odin::Costing::auto_shorter, "auto_shorter"},
      {odin::Costing::bicycle, "bicycle"},
      {odin::Costing::bus, "bus"},
      {odin::Costing::hov, "hov"},
      {odin::Costing::motor_scooter, "motor_scooter"},
      {odin::Costing::multimodal, "multimodal"},
      {odin::Costing::pedestrian, "pedestrian"},
      {odin::Costing::transit, "transit"},
      {odin::Costing::truck, "truck"},
      {odin::Costing::motorcycle, "motorcycle"},
      {odin::Costing::auto_data_fix, "auto_data_fix"},
  };
  auto i = costings.find(costing);
  return i == costings.cend() ? empty : i->second;
}
```

vs

```
static ::PROTOBUF_NAMESPACE_ID::internal::ExplicitlyConstructed<std::string> Costing_strings[12] = {};

static const char Costing_names[] =
  "auto_"
  "auto_data_fix"
  "auto_shorter"
  "bicycle"
  "bus"
  "hov"
  "motor_scooter"
  "motorcycle"
  "multimodal"
  "pedestrian"
  "transit"
  "truck";

static const ::PROTOBUF_NAMESPACE_ID::internal::EnumEntry Costing_entries[] = {
  { {Costing_names + 0, 5}, 0 },
  { {Costing_names + 5, 13}, 11 },
  { {Costing_names + 18, 12}, 1 },
  { {Costing_names + 30, 7}, 2 },
  { {Costing_names + 37, 3}, 3 },
  { {Costing_names + 40, 3}, 4 },
  { {Costing_names + 43, 13}, 5 },
  { {Costing_names + 56, 10}, 10 },
  { {Costing_names + 66, 10}, 6 },
  { {Costing_names + 76, 10}, 7 },
  { {Costing_names + 86, 7}, 8 },
  { {Costing_names + 93, 5}, 9 },
};

static const int Costing_entries_by_number[] = {
  0, // 0 -> auto_
  2, // 1 -> auto_shorter
  3, // 2 -> bicycle
  4, // 3 -> bus
  5, // 4 -> hov
  6, // 5 -> motor_scooter
  8, // 6 -> multimodal
  9, // 7 -> pedestrian
  10, // 8 -> transit
  11, // 9 -> truck
  7, // 10 -> motorcycle
  1, // 11 -> auto_data_fix
};

const std::string& Costing_Name(
    Costing value) {
  static const bool dummy =
      ::PROTOBUF_NAMESPACE_ID::internal::InitializeEnumStrings(
          Costing_entries,
          Costing_entries_by_number,
          12, Costing_strings);
  (void) dummy;
  int idx = ::PROTOBUF_NAMESPACE_ID::internal::LookUpEnumName(
      Costing_entries,
      Costing_entries_by_number,
      12, value);
  return idx == -1 ? ::PROTOBUF_NAMESPACE_ID::internal::GetEmptyString() :
                     Costing_strings[idx].get();
}

```
